File: core/src/ru/myitschool/dcrawler/dungeon/Door.java
package ru.myitschool.dcrawler.dungeon;

import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.utils.Array;

import java.awt.*;

/**
 * Created by Voyager on 09.05.2017.
 */
public class Door {

    private static Array<Door> doors = new Array<Door>();

    private int direction;

    Point[] doorCells;

    public Door(int direction, Point[] doorCells) {
        this.direction = direction;
        this.doorCells = doorCells;
        doors.add(this);
    }

    public int getDirection() {
        return direction;
    }
    
    public static int getDoorIndex(int x, int y){
        for (int i = 0; i < doors.size; i++) {
            Door door = doors.get(i);
            for (Point cell : door.doorCells) {
                if (cell.x == x && cell.y == y) {
                    return i;
                }
            }
        }
        return -1;
    }

    public static Door getDoor(int x, int y){
        return getDoor(getDoorIndex(x, y));
    }

    public static Door getDoor(int index){
        if (index != -1) {
            return doors.get(index);
        }
        return null;
    }

    public static void removeDoor(int x, int y){
        removeDoor(getDoorIndex(x, y));
    }

    public static void removeDoor(int index){
        if (index != -1) {
            doors.removeIndex(index);
        }
    }

    public Point[] getDoorCells() {
        return doorCells;
    }

    public void moveDoor(Vector2 tiles){
        for (Point cell : doorCells){
            cell.x += tiles.x;
            cell.y += tiles.y;
        }
    }

    public static void moveDoors(Vector2 tiles){
        for (Door door1 : doors){
            for (Point cell : door1.getDoorCells()){
                System.out.println("Door cell: (" + cell.x + "; " + cell.y + ")");
            }
        }
        System.out.println();
        for (Door door : doors){
            door.moveDoor(tiles);
        }
        for (Door door1 : doors){
            for (Point cell : door1.getDoorCells()){
                System.out.println("Door cell: (" + cell.x + "; " + cell.y + ")");
            }
        }
    }

    public static Array<Door> getDoors() {
        return doors;
    }
}

File: core/src/com/wolg_vlad/dcrawler/dungeon/Door.kt
package com.wolg_vlad.dcrawler.dungeon

import com.badlogic.gdx.math.Vector2
import java.awt.Point

/**
 * Created by Voyager on 09.05.2017.
 */
class Door(val direction: Int, var doorCells: Array<Point?>?) {

    fun moveDoor(tiles: Vector2) {
        for (cell in doorCells!!) {
            cell!!.x += tiles.x.toInt()
            cell!!.y += tiles.y.toInt()
        }
    }

    companion object {
        val doors = com.badlogic.gdx.utils.Array<Door>()
        @JvmStatic
        fun getDoorIndex(x: Int, y: Int): Int {
            for (i in 0 until doors.size) {
                val door = doors[i]
                for (cell in door.doorCells!!) {
                    if (cell!!.x == x && cell.y == y) {
                        return i
                    }
                }
            }
            return -1
        }

        fun getDoor(x: Int, y: Int): Door? {
            return getDoor(getDoorIndex(x, y))
        }

        @JvmStatic
        fun getDoor(index: Int): Door? {
            return if (index != -1) {
                doors[index]
            } else null
        }

        fun removeDoor(x: Int, y: Int) {
            removeDoor(getDoorIndex(x, y))
        }

        @JvmStatic
        fun removeDoor(index: Int) {
            if (index != -1) {
                doors.removeIndex(index)
            }
        }

        @JvmStatic
        fun moveDoors(tiles: Vector2) {
            for (door1 in doors) {
                for (cell in door1.doorCells!!) {
                    println("Door cell: (" + cell!!.x + "; " + cell.y + ")")
                }
            }
            println()
            for (door in doors) {
                door.moveDoor(tiles)
            }
            for (door1 in doors) {
                for (cell in door1.doorCells!!) {
                    println("Door cell: (" + cell!!.x + "; " + cell.y + ")")
                }
            }
        }

    }

    init {
        doors.add(this)
    }
}